private-network RFC (#67)
functionland/docs: docs/RFCs/private-network.md

- Feature Name: private-network
- Start Date: 2022-02-01
- RFC PR:
- Functionland Issue:

## Background

> we are using IPFS as our file system. but IPFS is build to use for public data, and it does not support ACL, 

Author: @gitaaron 
I don't think the security layer that cover all the stories would be ready in at least 3 month. and also i personally can't relay on access control by encryption only to make user safe. with this solution we can demo useful box and photo in in around a week.
As @masih mention i should explain more on public network risk's

> we are using IPFS as our file system. but IPFS is build to use for public data, and it does not support ACL, 

Reviewer: @farhoud  thanks for the clarification.  I am still not sure where this PR is coming from (probably because I joined the team after this work commenced).  Could you please share a few of the use cases / user stories this PR is solving?  Thanks!

> we are using IPFS as our file system. but IPFS is build to use for public data, and it does not support ACL, 

Reviewer: It seems to me that this PR is trying to cover two different scenarios - 

1.  Private communication between the fula-client and fula-api (eg/ uploading a photo to a BOX and retrieving a photo from a BOX)
2. Private communication between BOXes or other 'peers' in the fula-network (eg/ data syncing)

Is  that correct?

> we are using IPFS as our file system. but IPFS is build to use for public data, and it does not support ACL, 

Author: yes. and i try to add more on it.

so we need to find a way to keep user safe until our security layer become mature.

## Problem Statement
We need to protect user and his/her data from harms and risk of public network.  

## Motivation
Isolating User from public network can help us relax the problem and build useful product and testing our security layer without putting user in harms way. 

## Proposal
using IPFS build-in private network. base on [spec](https://github.com/libp2p/specs/blob/master/pnet/Private-Networks-PSK-V1.md)
Its uses a pre-shared key (PSK) to create a private network with encrypted communication.

> For box setup user provide an environment variable SECRET which is password he should remember.

Author: My English is a work on progress i will learn a lot from your comments.

the secret then convert to a hash of 256 bit by algorithm like sha256 and generate the swarm.key for ipfs and libp2p node's.

> For Fula client when user call createClient he should also provide the password he used for setting the box's.

Reviewer 2: Consider setting the scene before talking specifically about Fula client. e.g. a brief sentence reminding the reader what Fula client is and how the client operates in the private network, and whether its operation is any different.

```suggestion
For Fula client when user calls `createClient` they should also provide the password used for setting up the boxes.
```

* Use backtick for code and ideally link it up with API docs.
* Ditto re neutral pronouns.
* I recommend also setting the scene before this paragraph about the second secret. 

Question: Why two secrets at all? Why not use that secret for key generation as well?  The answer to these questions should be included in the doc.

> For Fula client when user call createClient he should also provide the password he used for setting the box's.

Author: I think by adding the human readable password i make this PR a little bit complicated. but my thinking was asking user to enter a 64  character in web app or photo's is painful task so i suggest to  change it to more human readable think.

> For Fula client when user call createClient he should also provide the password he used for setting the box's.

Author: I will try to separate them.

> For Fula client when user call createClient he should also provide the password he used for setting the box's.

Author: I think when we can use DID instead of Peer id to connect (verified connection) box <---> box and client <---> box which also most cover internal ipfs protocols like bitsawp engine then that's the end if life for this proposal

> For Fula client when user call createClient he should also provide the password he used for setting the box's.

Author: > I think when we can use DID instead of Peer id to connect (verified connection) box <---> box and client <---> box which also most cover internal IPFS like bitsawp engine then that's the end of life for this proposal

like the libp2p gatekeeper should use the DID to let others connect to my box or ... .

> For Fula client when user call createClient he should also provide the password he used for setting the box's.

Author: and if we want to use ipfs-cluster then it also have to cover a lot of ground on go lang too.

> For network discovery its manual process that user should provide all box peer id's in config.

Reviewer 2: > This RFC does not cover network discovery. instead it is assumed that each peer is configured with the list of all other peer IDs in the private network.

Consider moving this to Scope of work.

> 

Reviewer 2: We also need to talk about how the network is joined together. So far this reads as a simple key generation RFC.

Author: I will try to add current network topology to background and also how network join in the proposal

## Scope of work

> - human friendly password to swarm.key 

Author: duplicated

> - human friendly password to swarm.key 

Author: I don't see any overlapping. we are working at different layer.

- box should look for SECRET env variable and if it exit set the connProtector for libp2p config
- fula-client constructor should get another parameter call secret and if exist set connProtector for libp2p

> - create an example for the describing functionality

Reviewer 2: For me the items listed here sound like "Actionables" more than scope of work. I'd expect a section titled as such to tell the reader to what extent the solutions documented by this RFC solve the bigger problem, along with list of things that are _not_ solved or any simplifying assumptions made.

> ## Implementation 

Reviewer 2: Consider providing an overview of steps, where each sub section expands on.

> ### human-readable password

Reviewer 2: I am curious what is special about the human readability of password in the context of this document? If it is significant in the context of private network to have human readable password then we need a section justifying the need for it. 

Otherwise, my recommendation is to simplify.

Convert a human-readable password to swarm.key
simplest way is to use ['libp2p-crypto'](https://github.com/libp2p/js-libp2p-crypto#cryptopbkdf2password-salt-iterations-keysize-hash) (it works both on browser and node)
```js
import crypto from 'libp2p-crypto'
const salt = '';
const passwordToPKey = (password) => {
  // it should be somthing like this
  const key = crypto.pbkdf2(password, salt, 5, 256, 'sha2-256')
  pkey = `/key/swarm/psk/1.0.0/
  /base16/
  ${key}`
  return new TextEncoder().encode(key)
}

```

> the box and client already support pkey input, but we should add the above function to them

Reviewer 2: * Capitalise first letter in paragraphs. I'll stop pointing this out for the rest of the document.

* Seems like there is a specific place in code you envision this needs to be added? If so, link it up.

and after that we should change existing pkey on both fula and box to use the above function.

> for creating an example we can create copy of react-gallery and add the password field in config.<br/>

Reviewer 2: Flesh this out as a case study with more details. Remove `<br/>`?

note: if example repo would be outside mono-repo we can just use branch for describing every functionality.

## Alternative approaches

> - Using VPN for creating the private network

Reviewer 2: Expand on this please.

Author: @gitaaron
Can you collaborate with me on this?

Reviewer: Apologies I think I am mistaken on the use case / purpose of this PR.  You can remove this.

>   - It adds another point of failure to the system.

Reviewer 2: How?

Author: @gitaaron @ehsan6sha 
I just try to reflect the conversion that Ehsan and Aaron had in issue's
@gitaaron 
Can you collaborate with me on this?
